sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java

```diff
@@ -79,24 +83,45 @@ public class ChainedTokenCredential implements TokenCredential {
-        return Flux.fromIterable(credentials)
-            .flatMap(p -> p.getToken(request)
-                .doOnNext(t -> LOGGER.info("Azure Identity => Attempted credential {} returns a token",
-                    p.getClass().getSimpleName()))
+        Mono<AccessToken> accessTokenMono;
+        if (cachedWorkingCredential.get() != null) {
+            accessTokenMono =  Mono.defer(() -> cachedWorkingCredential.get().getToken(request)
+                .doOnNext(t -> LOGGER.info("Azure Identity => Returning token from cached credential {}",
+                    cachedWorkingCredential.get().getClass().getSimpleName()))
                 .onErrorResume(Exception.class, t -> {
                     if (!t.getClass().getSimpleName().equals("CredentialUnavailableException")) {
                         return Mono.error(new ClientAuthenticationException(
-                            unavailableError + p.getClass().getSimpleName()
+                            unavailableError + cachedWorkingCredential.get().getClass().getSimpleName()
                                 + " authentication failed. Error Details: " + t.getMessage(),
                             null, t));
                     }
                     exceptions.add((CredentialUnavailableException) t);
-                    LOGGER.info("Azure Identity => Attempted credential {} is unavailable.",
-                        p.getClass().getSimpleName());
+                    LOGGER.info("Azure Identity => Cached credential {} is unavailable.",
+                        cachedWorkingCredential.get().getClass().getSimpleName());
                     return Mono.empty();
-                }), 1)
-            .next()
-            .switchIfEmpty(Mono.defer(() -> {
+                }));
+        } else {
+            accessTokenMono = Flux.fromIterable(credentials)
+                .flatMap(p -> p.getToken(request)
+                    .doOnNext(t -> {
+                        LOGGER.info("Azure Identity => Attempted credential {} returns a token",
+                            p.getClass().getSimpleName());
+                        cachedWorkingCredential.set(p);
+                    }).onErrorResume(Exception.class, t -> {
+                        if (!t.getClass().getSimpleName().equals("CredentialUnavailableException")) {
+                            return Mono.error(new ClientAuthenticationException(
+                                unavailableError + p.getClass().getSimpleName()
+                                    + " authentication failed. Error Details: " + t.getMessage(),
+                                null, t));
+                        }
+                        exceptions.add((CredentialUnavailableException) t);
+                        LOGGER.info("Azure Identity => Attempted credential {} is unavailable.",
+                            p.getClass().getSimpleName());
+                        return Mono.empty();
+                    }), 1)
+                .next();
```

Comment:
Can we factor this into a method so we don't repeat all the logic? Could just have an extra logger in the cache case to emit an extra line that says "using cache" or the like.

Reply:
Yeah, agreed.
Already have a change for it.